### PR TITLE
Attempt to add pagination functions

### PR DESCRIPTION
**Pull request by keyneston**
This is an experiment at adding paging functions for querying. It is similar to what I've seen from AWS and similar libraries. It just abstracts the looping and what not from you.

Example:
```go
	results := []*chat.Channel{}

	err = client.QueryChannelsPages(&chat.QueryOption{
		Filter: map[string]interface{}{
			"members": map[string]interface{}{"$in": []string{"555"}},
		},
	}, func(more []*chat.Channel) bool {
		results = append(results, more...)
		return true
	})

	if err != nil {
		panic(err)
	}
```

Thoughts on the idea, and style? It already is helping me with some of the scripts I'm writing for support purposes.

### CODE REVIEW

File: query.go [hunk 1]
```diff
@@ -133,6 +134,38 @@ func (c *Client) QueryChannels(q *QueryOption, sort ...*SortOption) ([]*Channel,
 	return result, err
 }
 
+type ChannelPaginationFunc func([]*Channel) bool
+
+func (c *Client) QueryChannelsPages(q *QueryOption, paginationFunc ChannelPaginationFunc) error {
+	if q == nil {
+		return fmt.Errorf("what should be done here?")
```
Comment (ferhatelmas) [inline query.go:141]: defaults, IMHO

File: query.go [hunk 2]
```diff
@@ -133,6 +134,38 @@ func (c *Client) QueryChannels(q *QueryOption, sort ...*SortOption) ([]*Channel,
...
+	}
+
+	// Make a copy of q so we can manipulate it without affecting memory outside of here
+	newQ := QueryOption{}
+	newQ = *q
```
Comment (ferhatelmas) [inline query.go:146]: why not `newQ := *q` directly?

File: query.go [hunk 3]
```diff
@@ -133,6 +134,38 @@ func (c *Client) QueryChannels(q *QueryOption, sort ...*SortOption) ([]*Channel,
...
+
+	limit := 30
```
Comment (ferhatelmas) [inline query.go:148]: it should care given limit maybe

File: query.go [hunk 4]
```diff
@@ -133,6 +134,38 @@ func (c *Client) QueryChannels(q *QueryOption, sort ...*SortOption) ([]*Channel,
...
+
+	for i := 0; ; i++ {
+		newQ.Limit = limit
+		newQ.Offset = limit * i
```
Comment (ferhatelmas) [inline query.go:152]: also caring for the starting offset, maybe

**Comment by ferhatelmas** [comment]
we might alias it to `type M = map[string]interface{}`

**Comment by ferhatelmas** [comment]
For naming, `client.PageQueryChannels` sounds better to me than `client.QueryChannelsPages`.

**Comment by keyneston** [comment]
@ferhatelmas thanks for the feedback!

> For naming, client.PageQueryChannels sounds better to me than client.QueryChannelsPages.

Definitely agree.

> we might alias it to type M = map[string]interface{} to make it shorter.

In scripty code I often do `type obj map[string]interface{}`, I tend to keep libraries, prod code, and etc to the full name. That said I do really like `logrus.Fields` which is just an alias. Maybe:

```go
type Filter map[string]interface{}
```
Or something?

**Comment by ferhatelmas** [comment]
Very welcome. A bit more context for the alias suggestion.

My inspiration for alias comes from web frameworks:

[echo has Map](https://godoc.org/github.com/labstack/echo#Map)(was M before), [gin has H](https://godoc.org/github.com/gin-gonic/gin#H) for the same type.

Since it's common for a JSON response, they wanted to make it shorter for the repeated pattern I guess.

But in their case, it's not an alias since they need to attach some methods for different kinds of serialization.

**Pull request closed without merging**